Task1.Generate a reproducible sensor matrix
Create a 2D NumPy array named readings with shape (360, 4) to represent six hours of minute-level readings from four sensors. Use a fixed seed and generate values that roughly follow a normal distribution around 50 with some variation. After creating the array, print its shape, dtype, and the first three rows to confirm the structure.

In [25]:
import numpy as np
np.random.seed(42)

In [26]:
rows=360
cols=4
mean = 50       
std_dev = 15      
readings = np.random.normal(loc=mean, scale=std_dev, size=(rows, cols))


In [27]:
print("Shape:", readings.shape)      
print("Dtype:", readings.dtype)      
print("First 3 rows:\n", readings[:3]) 

Shape: (360, 4)
Dtype: float64
First 3 rows:
 [[57.4507123  47.92603548 59.71532807 72.84544785]
 [46.48769938 46.48794565 73.68819223 61.51152094]
 [42.95788421 58.13840065 43.04873461 43.0140537 ]]


Shape: (360, 4)
Dtype: float64
First 3 rows:
 [[51.25730221 48.67895137 56.4042265  51.04900117]
 [44.64330627 53.61595055 63.04000045 59.47080963]
 [42.96264764 37.34578529 43.76725537 50.41325979]]


#### Task 3: Filter outliers with boolean masks
Define outliers as any reading above 80 or below 20. Build a boolean mask that identifies outliers across the full matrix. Use the mask to compute two values: the count of outlier readings and the percentage of total readings that are outliers. Then create a cleaned array cleaned_readings where outliers are replaced with np.nan. Confirm that the number of np.nan values matches your outlier count.

In [28]:
outlier_mask = (readings > 80) | (readings < 20)

In [29]:
outlier_count = np.sum(outlier_mask)
print("Outlier count:", outlier_count)

Outlier count: 65


Outlier count: 56


In [30]:
total_readings=readings.size
outlier_percentage=(outlier_count/total_readings)*100
print("Outlier percentage:", outlier_percentage, "%")

Outlier percentage: 4.513888888888888 %


In [31]:
cleaned_readings = readings.copy()
cleaned_readings[outlier_mask] = np.nan

In [32]:
nan_count = np.isnan(cleaned_readings).sum()
print("NaN count:", nan_count)
print("Counts match:", nan_count == outlier_count)

NaN count: 65
Counts match: True


NaN count: 56
Counts match: True

#### Task 4: Compute summaries by sensor and over time
Compute the mean and standard deviation for each sensor using axis-based aggregation. Store the results in two 1D arrays named sensor_means and sensor_stds. Then compute the average reading at each time step across sensors and store it in a 1D array named time_means. Validate shapes: the sensor summaries should have length 4, and the time summary should have length 360.

In [33]:
sensor_means = np.mean(readings, axis=0)
sensor_stds = np.std(readings, axis=0)

print("Sensor means:", sensor_means)
print("Sensor stds:", sensor_stds)

Sensor means: [50.43849455 50.22541801 50.83432074 50.89241994]
Sensor stds: [13.9980601  15.41966798 14.79864467 14.95712768]


Sensor means: [50.2921572  51.68408191 50.0236014  49.59378015]
Sensor stds: [14.01070979 14.17641004 15.170803   15.14529985]

In [34]:
time_means = np.mean(readings, axis=1)
print("Time means shape:", time_means.shape)

Time means shape: (360,)


In [35]:
print("Sensor means shape:", sensor_means.shape)
print("Sensor stds shape:", sensor_stds.shape)
print("Time means shape:", time_means.shape)

Sensor means shape: (4,)
Sensor stds shape: (4,)
Time means shape: (360,)


#### Task 5: Build a validation report (Part 1)
Create a short report dictionary with keys total_readings, outlier_count, outlier_percent, sensor_means, and sensor_stds. Convert this report into a readable string and print it in the notebook. Include at least one simple validation, such as confirming that total_readings equals readings.size.

In [36]:
report = {
    "total_readings": readings.size,
    "outlier_count": outlier_count,
    "outlier_percent": outlier_percentage,
    "sensor_means": sensor_means,
    "sensor_stds": sensor_stds
}

In [37]:
validation_check = (report["total_readings"] == readings.size)
report["size_validation"] = validation_check
print(validation_check)

True


In [43]:
report_str=f"""
Total readings: {report['total_readings']}
Outlier count: {report['outlier_count']}
Outlier percentage: {report['outlier_percent']:.2f} %

Sensor means: {report['sensor_means']}
Sensor std deviations: {report['sensor_stds']}

Size validation passed: {report['size_validation']}
"""
print(report_str)


Total readings: 1440
Outlier count: 65
Outlier percentage: 4.51 %

Sensor means: [50.43849455 50.22541801 50.83432074 50.89241994]
Sensor std deviations: [13.9980601  15.41966798 14.79864467 14.95712768]

Size validation passed: True



Total readings: 1440
Outlier count: 65
Outlier percentage: 4.51 %

Sensor means: [50.43849455 50.22541801 50.83432074 50.89241994]
Sensor std deviations: [13.9980601  15.41966798 14.79864467 14.95712768]

Size validation passed: True

#### Task 2: Apply vectorized transformations
Create a new array scaled_readings that rescales the original readings to a 0–1 range using vectorized operations. Do not use loops. Then create a second array centered_readings by subtracting the column means from the original readings. Verify both arrays have the same shape as the original.

In [46]:
min_val = np.min(readings)
max_val = np.max(readings)
scaled_readings = (readings - min_val) / (max_val - min_val)
scaled_readings.shape

(360, 4)

In [47]:
column_means = np.mean(readings, axis=0)
centered_readings = readings - column_means
centered_readings.shape

(360, 4)

In [48]:
print("Original shape:", readings.shape)
print("Scaled shape:", scaled_readings.shape)
print("Centered shape:", centered_readings.shape)

Original shape: (360, 4)
Scaled shape: (360, 4)
Centered shape: (360, 4)


Original shape: (360, 4)
Scaled shape: (360, 4)
Centered shape: (360, 4)

### Part 2: Simulation, Broadcasting, and Randomness
#### Task 6: Create a reproducible simulation matrix
Initialize a random number generator with a fixed seed. Create a 2D array named sim with shape (1000, 6) representing 1000 trials for 6 scenarios. Use a normal distribution with mean 0 and standard deviation 1. After creating the array, print its shape, dtype, and the first three rows to confirm the structure.

In [10]:
import numpy as np
rng=np.random.default_rng(seed=77)

In [11]:
sim = rng.normal(loc=0, scale=1, size=(1000,6)) 

In [12]:
print("Shape:", sim.shape)
print("Data type:", sim.dtype)
print("First 3 rows:\n", sim[:3])

Shape: (1000, 6)
Data type: float64
First 3 rows:
 [[ 0.42776996 -0.57083756  2.65446069 -1.60854495  0.66171566 -0.14342594]
 [-0.35450639  1.06635881 -1.817922   -0.98467621 -0.11416014  1.74127384]
 [ 0.08904687  0.89568824 -1.86330597 -1.23888755  0.96952947 -0.62817974]]


Shape: (1000, 6)
Data type: float64
First 3 rows:
 [[ 0.42776996 -0.57083756  2.65446069 -1.60854495  0.66171566 -0.14342594]
 [-0.35450639  1.06635881 -1.817922   -0.98467621 -0.11416014  1.74127384]
 [ 0.08904687  0.89568824 -1.86330597 -1.23888755  0.96952947 -0.62817974]]

#### Task 7: Apply broadcasting for scenario adjustments
Create a 1D array named scenario_shift of length 6 containing small offsets such as [-0.3, -0.1, 0.0, 0.1, 0.2, 0.4]. Use broadcasting to add these offsets to sim and store the result in adjusted_sim. Verify that adjusted_sim has the same shape as sim and that the mean of each column changes in the direction of the offsets.

In [13]:
scenario_shift = np.array([-0.3, -0.1, 0.0, 0.1, 0.2, 0.4])
print(scenario_shift.shape)

(6,)


In [14]:
adjusted_sim = sim + scenario_shift

In [15]:
sim.shape,adjusted_sim.shape

((1000, 6), (1000, 6))

In [16]:
original_means = np.mean(sim, axis=0)
adjusted_means = np.mean(adjusted_sim, axis=0)
original_means,adjusted_means

(array([-0.03929517,  0.03371686,  0.02442738, -0.0227731 , -0.03133955,
         0.04724679]),
 array([-0.33929517, -0.06628314,  0.02442738,  0.0772269 ,  0.16866045,
         0.44724679]))

(array([-0.03929517,  0.03371686,  0.02442738, -0.0227731 , -0.03133955,
         0.04724679]),
 array([-0.33929517, -0.06628314,  0.02442738,  0.0772269 ,  0.16866045,
         0.44724679]))

In [17]:
mean_difference = adjusted_means - original_means
mean_difference

array([-0.3, -0.1,  0. ,  0.1,  0.2,  0.4])

In [ ]:
array([-0.3, -0.1,  0. ,  0.1,  0.2,  0.4])


#### Task 8: Rank scenarios with sorting and partitioning
Compute the mean outcome per scenario from adjusted_sim and store it in a 1D array named scenario_means. Use np.argsort to get the ranking of scenarios from lowest to highest. Then use np.partition to identify the top two scenario means without fully sorting the array. Confirm that the top two values from partition match the two largest values from a full sort.

In [21]:
scenario_means = np.mean(adjusted_sim, axis=0)
scenario_means

array([-0.33929517, -0.06628314,  0.02442738,  0.0772269 ,  0.16866045,
        0.44724679])

In [22]:
ranking_indices=np.argsort(scenario_means)
print("Ranking of scenarios (lowest to highest):", ranking_indices)

Ranking of scenarios (lowest to highest): [0 1 2 3 4 5]


Ranking of scenarios (lowest to highest): [0 1 2 3 4 5]


In [23]:
top2_partition = np.partition(scenario_means, -2)[-2:]
top2_partition

array([0.16866045, 0.44724679])

array([0.16866045, 0.44724679])


#### Task 9: Controlled randomness and reproducibility 
Generate a second simulation matrix sim_2 using the same seed and confirm that it matches sim exactly. Then generate a third matrix sim_3 with a different seed and confirm that it differs. Include a short check such as comparing equality counts or using np.allclose to verify the difference.

In [38]:
rng2 = np.random.default_rng(seed=77)
sim_2 = rng2.normal(loc=0, scale=1, size=(1000, 6))
print("sim and sim_2 are exactly equal:", np.array_equal(sim, sim_2))

sim and sim_2 are exactly equal: True


sim and sim_2 are exactly equal: True


In [39]:
rng3 = np.random.default_rng(seed=42)
sim_3 = rng3.normal(loc=0, scale=1, size=(1000, 6))
print("sim and sim_3 are exactly equal:", np.array_equal(sim, sim_3))
print("sim and sim_3 are close:", np.allclose(sim, sim_3))

sim and sim_3 are exactly equal: False
sim and sim_3 are close: False


sim and sim_3 are exactly equal: False
sim and sim_3 are close: False


#### Task 10: Build a validation report (Part 2)
Create a small report dictionary with keys shape, scenario_means, top_two_indices, and top_two_values. Convert this report into a readable string and print it. Include at least one validation statement, such as confirming that scenario_means has length 6 and that the top_two_indices list has length 2.



In [42]:
report2 = {
    "shape": adjusted_sim.shape,
    "scenario_means": scenario_means,
    "top_two_indices": ranking_indices[-2:],
    "top_two_values": top2_partition        
}

In [43]:
report_str = f"""
Simulation Report (Part 2)
Matrix shape: {report2['shape']}
Scenario means: {report2['scenario_means']}
Top two scenario indices: {report2['top_two_indices']}
Top two scenario values: {report2['top_two_values']}
"""

print(report_str)


Simulation Report (Part 2)
Matrix shape: (1000, 6)
Scenario means: [-0.33929517 -0.06628314  0.02442738  0.0772269   0.16866045  0.44724679]
Top two scenario indices: [4 5]
Top two scenario values: [0.16866045 0.44724679]



Simulation Report (Part 2)
Matrix shape: (1000, 6)
Scenario means: [-0.33929517 -0.06628314  0.02442738  0.0772269   0.16866045  0.44724679]
Top two scenario indices: [4 5]
Top two scenario values: [0.16866045 0.44724679]